homeassistant/components/mqtt/__init__.py

```diff
@@ -577,7 +578,7 @@ def remove():
     success = await hass.data[DATA_MQTT].async_connect()  # type: bool
 
     if not success:
-        return False
+        raise ConfigEntryNotReady
```

Comment:
We should make sure that we don't throw this exception when the error is not recoverable, like password is wrong.

Reply:
That's why I was thinking about throwing it here originally: https://github.com/home-assistant/home-assistant/blob/40170ffc04546b8eb88897b6f7c12b5c0054f2c9/homeassistant/components/mqtt/__init__.py#L705

But I'm not sure if `OSError` overlaps with the non-recoverable situation.

Reply:
Well since we control `async_connect`, we can have it return a constant. And we can have 3 constants: `CONNECTION_SUCCESS`, `CONNECTION_FAILED`, `CONNECTION_FAILED_RECOVERABLE`.

Reply:
How does it look now?

Ignoring the failing async test that I don't understand.